### ISSUE TITLE
support chloropleth maps with colormaps

### ISSUE THREAD

**Issue by yeesian** [yeesian/GeoDataFrames.jl#1]
@tbreloff following up from https://github.com/JuliaGeo/Shapefile.jl/pull/10, I currently have the following going:

![image](https://cloud.githubusercontent.com/assets/1231079/18520017/0e0f1d94-7a74-11e6-8e72-3f272283db23.png)

with code that's similar to https://github.com/yeesian/GeoDataFrames.jl/blob/166338156ade81727d186cfe7a07c8d03eac12fd/src/GeoDataFrames.jl#L25-L37:

``` julia
function plot(df::DataFrame; plt = Plots.plot(bg = :black), geom::Symbol=:none, label::Symbol=:none, kwargs...)
    geometries = (geom == :none) ? df[:geometry0] : df[geom]
    for (i,geom) in enumerate(geometries)
        DataFrames.isna(geom) && continue
        if label != :none
            plot!(plt, shape(geom); label=df[i,label], kwargs...)
        else
            plot!(plt, shape(geom); kwargs...)
        end
    end
    plt
end
```

But I'm struggling to follow the development in Plots (or PlotRecipes or RecipesBase) to figure out what needs to be done to support chloropleth maps. Might you have any pointers for me to look through? Feedback on better practices will be welcome too!

@mkborregaard: feel free to chip in

**Comment by tbreloff** [comment]
> to figure out what needs to be done to support chloropleth maps

Looks like it's already working for you?  I'm not sure exactly what you need to figure out

> Feedback on better practices will be welcome too

There's a good chance that this could depend on RecipesBase and you could write a "type recipe" to convert an `AbstractVector{Geometry}` to vectors of x/y coords.  Then most of the plotting code goes away.  See https://juliaplots.github.io/recipes/

**Comment by yeesian** [comment]
> Looks like it's already working for you?

It's currently assigning a random color to each shape/polygon. [Chloropleth maps](https://en.wikipedia.org/wiki/Choropleth_map) usually color them according to some [user-specified] column of attribute values, based on a default/user-specified [colormap](https://github.com/tbreloff/ExamplePlots.jl/blob/master/notebooks/scratch/colormaps.ipynb).

**Comment by tbreloff** [comment]
Gotcha... the answer will be to use something like `marker_z`.  I'm not sure if it's implemented for shapes though... I'd need to check.

**Comment by tbreloff** [comment]
You have to set the colors yourself, but it's not too hard:

``` julia
using Plots
shapes = [Shape(rand(3),rand(3)) for i=1:10]
z = 1:10
grad = cgrad()
plot(shapes, fc=[grad[zi/10] for zi=z])
```

![tmp](https://cloud.githubusercontent.com/assets/933338/18521156/0840b0f8-7a79-11e6-929e-35fc021214f6.png)

**Comment by yeesian** [comment]
That's what I was looking for, thanks!

> You have to set the colors yourself

I take it there isn't yet a straightforward way to do it using type recipes?

**Comment by yeesian** [comment]
Reopening the issue until it's pushed to master

**Comment by tbreloff** [comment]
Well, the type recipe would convert your Geometries to Shapes (or just the coordinates).  Then you could separately define the colors (this could be done inside the recipe, of course, if you have all the data you need available)

**Comment by yeesian** [comment]
gotcha

**Comment by tbreloff** [comment]
It's preferable to let the user specify colors directly if you can, which means you wouldn't do anything except the "Geometry --> coords" conversion.

**Comment by mkborregaard** [comment]
Isn't the Geometries-->coords conversion exactly what is already in the Shapefile recipe in PlotRecipes?

**Comment by yeesian** [comment]
> Isn't the Geometries-->coords conversion exactly what is already in the Shapefile recipe in PlotRecipes?

Yup, it is, but for OGRGeometries (which has drivers for a [much wider range of file formats](http://www.gdal.org/ogr_formats.html)). This was just a quick hack from me to learn about the various components, as a quick demo for http://www.meetup.com/Maptime-Boston/events/234117226/; it probably isn't going to be the final form of the package.

**Comment by tbreloff** [comment]
If that's true (sorry I don't remember) then you shouldn't have to do anything.  It should "just work"

**Comment by mkborregaard** [comment]
We have only implemented it for a single type of shapefile (simple polygons). Lines and points are missing. Complex polygons (i.e. polygons with inner holes) may require some careful thinking? I am happy to do the missing recipes soon if they are needed.

**Comment by yeesian** [comment]
> I am happy to do the missing recipes soon if they are needed.

I'll be happy to take PRs!

**Comment by yeesian** [comment]
closed by #3
